build.sbt

```diff
@@ -154,6 +153,15 @@ Compile / doc / scalacOptions ++= Seq(
 )
 
 assembly / assemblyJarName := s"${name.value}-assembly-${sparkVerStr.value}_${scalaBinaryVersion.value}-${version.value}.jar"
+// Excluding Scala library jars, see https://github.com/sbt/sbt-assembly/tree/v1.2.0#excluding-scala-library-jars
+assemblyPackageScala / assembleArtifact := false
+assembly / assemblyExcludedJars := {
+  val cp = (assembly / fullClasspath).value
+  cp filter { x =>
```

Comment:
Is it possible to change the assembly inventory by changing the configurations of a dependency rather than filter by name?
Can we try the configs in these two files?
https://github.com/sbt/librarymanagement/blob/a83f75c02e91afa114c77dae1b5a8f084913b1e4/core/src/main/scala/sbt/librarymanagement/LibraryManagementSyntax.scala#L28-L36

https://github.com/sbt/librarymanagement/blob/a83f75c02e91afa114c77dae1b5a8f084913b1e4/core/src/main/scala/sbt/librarymanagement/Configuration.scala#L46-L58

------
I was a user of `gradle`, in `gradle`, we can set a dependency to `compile`(which is actually `compile` and `runtime`), or `compileOnly`, if it's `compileOnly`, the assembly plugin will not include it.

Reply:
I'm not sure, but I guess `Compile` in `sbt` is `compileOnly` in `gradle`, and `Runtime` in `sbt` is `Compile` in `gradle`.

Reply:
Just mark `enableif` as Provided, it will not be assembled. But I think it is better to exclude it explicitly. Because I think `Provided` is semanticly wrong.

Reply:
And antlr4-runtime could be marked as Provided, because it is truely provided via Apache Spark.

Reply:
Well, `antlr4-runtime` is brought by the sbt plugin, the `antlr4-runtime` brought by spark deps is already `Provided`.

Since `antlr4-runtime` has zero dependency, I think it is ok to exclude it by classpath filter.

Reply:
Well, `antlr4-runtime` is brought by the sbt plugin, the `antlr4-runtime` brought by spark deps is already `Provided`.

Since `antlr4-runtime` has zero dependency, I think it is ok to exclude it by classpath filter.

Reply:
I don't think `Provided` is semantically wrong, from the definition [but indicates you expect the JDK or a container to provide the dependency at runtime](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope). Here in our current scenario, the runtime of `enableif` is project build time, it do provide the dependency.

Reply:
I don't think `Provided` is semantically wrong, from the definition [but indicates you expect the JDK or a container to provide the dependency at runtime](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope). Here in our current scenario, the runtime of `enableif` is project build time, it do provide the dependency.

Reply:
I have a further idea, maybe we can make `enableif` an sbt plugin, like put `addSbtPlugin("com.thoughtworks.enableIf" % "enableif" % "1.1.8")` into `plugins.sbt`, then we don't have this bothering. But I just tried it, it failed, maybe we can do some additional work to make it a plugin.

Reply:
See https://stackoverflow.com/questions/35477974/add-a-compile-time-only-sub-project-dependency-in-sbt/35486527

We could use CompileOnly

Reply:
I googled about it when introducing enableif.scala

I though it would increase the complexity of the sbt build def.

Reply:
`compile-internal` is the solution for enableif.scala